packages/@react-spectrum/utils/src/BreakpointProvider.tsx

```diff
@@ -0,0 +1,86 @@
+import React, {ReactNode, useContext, useEffect, useState} from 'react';
+import {useIsSSR} from '@react-aria/ssr';
+
+interface Breakpoints {
+  S?: number,
+  M?: number,
+  L?: number,
+  [custom: string]: number
+}
+
+interface IBreakpointContext {
+  matchedBreakpoints: string[]
+}
+
+export const BreakpointContext = React.createContext<IBreakpointContext>(null);
+BreakpointContext.displayName = 'BreakpointContext';
+
+// This is temporary values.
+export const DEFAULT_BREAKPOINTS = {S: 380, M: 768, L: 1024};
+
+interface BreakpointProviderProps {
+  children?: ReactNode,
+  matchedBreakpoints: string[]
+}
+
+export function BreakpointProvider(props: BreakpointProviderProps) {
+  let {
+    children,
+    matchedBreakpoints
+  } = props;
+  return (
+    <BreakpointContext.Provider
+      value={{matchedBreakpoints}} >
+      {children}
+    </BreakpointContext.Provider>
+  );
+}
+
+export function useMatchedBreakpoints(breakpoints: Breakpoints, mobileFirst: boolean): string[] {
+  // sort breakpoints
+  let entries = Object.entries(breakpoints).sort(([, valueA], [, valueB]) => mobileFirst ? valueB - valueA : valueA - valueB);
+  let breakpointQueries = entries.map(([, value]) => mobileFirst ?  `(min-width: ${value}px)` : `(max-width: ${value}px)`);
```

Comment:
I think this might be in reverse?
in mobileFirst, you get base at small screen sizes
and in mobileFirst=false, you get base at large screen sizes
I think base would usually be a kind of middle ground? so in mobile first we wouldn't ever want base to show up on small screen, only as we get to larger screens?

I'm not sure, @devongovett @LFDanLu do either of you have an opinion on this?

Reply:
I changed the code a lot and it might have confused you, sorry.
Yeah, I reversed this for `getResponsiveProp` function.

```ts
export function getResponsiveProp<T>(prop: Responsive<T>, matchedBreakpoints: Breakpoint[]): T {
  if (prop && typeof prop === 'object' && !Array.isArray(prop)) {
    for (let i = 0; i < matchedBreakpoints.length; i++) {
      let breakpoint = matchedBreakpoints[i];
      if (prop[breakpoint]) {
        return prop[breakpoint];
      }
    }
    return (prop as ResponsiveProp<T>).base;
  }
  return prop as T;
}
```

> in mobileFirst, you get base at small screen sizes
and in mobileFirst=false, you get base at large screen sizes

Yes, I think when we design CSS for the mobile-first strategy, we write mobile CSS code as a base, and write CSS for larger scopes after that. LIke:

```scss
.class {
  // this is for mobile
  font-size: 20px;

  // this is for tablet
  @media screen and (min-width: 480px) {
    font-size: 18px;
  }

  // this is for desktop
  @media screen and (min-width: 1024px) {
    font-size: 16px;
  }
}
```

Same to the case for `mobileFirst = false` which means desktop-first.

```scss
.class {
  // this is for desktop
  font-size: 16px;

  // this is for tablet
  @media screen and (max-width: 1024px) {
    font-size: 18px;
  }

  // this is for mobile
  @media screen and (max-width: 480px) {
    font-size: 20px;
  }
}
```

So I thought base must be matched on every screen sizes.

Reply:
that makes sense now, sorry about that confusion